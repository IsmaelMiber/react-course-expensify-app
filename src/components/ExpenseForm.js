import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createAt: this.props.expense ? moment(this.props.expense.createAt) : moment(),
        calendarFocused: false,
        error: '',
    };

    onDescriptionChange = (e) => {
        const description = e.target.value.trim();
        this.setState(() => ({description}));
    }

    onNoteChange = (e) => {
        const note = e.target.value.trim();
        this.setState(() => ({note}));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ) this.setState(() => ({amount}));
    }

    onDateChange = (createAt) => {
        if(createAt) this.setState(() => ({createAt}));
    }

    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused: focused}));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'Please, provide description & amount'}));
        } else {
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createAt: this.state.createAt.valueOf(),
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error ? <p>{this.state.error}</p> : this.state.error}

                <form onSubmit={this.onSubmit}>
                    <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    />

                    <input
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    />

                    <textarea
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Submit</button>
                </form>
            </div>
        );
    }
}