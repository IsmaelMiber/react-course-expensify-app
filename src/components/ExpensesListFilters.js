import React from 'react';
import { connect } from 'react-redux';
import * as filterActions from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpensesListFilters extends React.Component {

    state = {
        focusedInput: null,
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (focusedInput) => this.setState(()=>({focusedInput}));

    onSortByChange = (e) => e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();

    onTextChange =  (e) => this.props.setTextFilter(e.target.value);

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                 />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}

                />
            </div>
        );
    }
}

const mapStateToProps = (state) => (
    {
        filters: state.filters,
    }
);

export default connect(mapStateToProps, filterActions)(ExpensesListFilters);