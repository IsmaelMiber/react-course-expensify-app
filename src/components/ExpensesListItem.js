import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpensesListItem = (props) => (
    <div>
        <p><Link to={`/edit/${props.id}`}>description: {props.description}</Link></p>
        <p>amount: {numeral(props.amount / 100).format("$0,0.00")}</p>
        <p>createdAt: {moment(props.createAt).format("MMMM Do, YYYY")}</p>
        <hr />
    </div>
);

export default ExpensesListItem;