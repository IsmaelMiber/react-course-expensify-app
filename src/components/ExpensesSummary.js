import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expenses-total';

export const ExpensesSummary = props => (
    <p>
        Viewing {props.expensesCount} expenses totalling {numeral(props.expensesTotal / 100).format("$0,0.00")}
    </p>
);

const mapStateToProps = state => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: totalExpenses(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);