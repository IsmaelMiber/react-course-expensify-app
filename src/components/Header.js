import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <ul>
            <li> <NavLink to="/" activeClassName="is-active" exact={true} >Home</NavLink> </li>
            <li> <NavLink to="/create" activeClassName="is-active" exact={true} >Create Expense Page</NavLink> </li>
            <li> <NavLink to="/help" activeClassName="is-active" exact={true} >Help Page</NavLink> </li>
        </ul>
    </header>
);

export default Header;