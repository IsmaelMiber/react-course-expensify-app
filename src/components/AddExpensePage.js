import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);