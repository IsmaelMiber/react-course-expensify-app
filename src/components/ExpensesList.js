import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem  from './ExpensesListItem';
import selectExpenses from '../selectors/expenses';

export const ExpensesList = (props) => (
    <div>
        <h1>Expenses List</h1>
        { props.expenses.length !== 0 ?
            props.expenses.length > 0 && props.expenses.map(expense => (
            <ExpensesListItem key={expense.id} {...expense}/>
        )) : (
             <p>There is no expenses</p>
             )
        }
    </div>
);

const mapStateToProps = state => (
    {
        expenses: selectExpenses(state.expenses, state.filters),
    }
);

export default connect(mapStateToProps)(ExpensesList);