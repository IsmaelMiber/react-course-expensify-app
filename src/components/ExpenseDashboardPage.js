import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesListFilters from './ExpensesListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpensesListFilters />
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;