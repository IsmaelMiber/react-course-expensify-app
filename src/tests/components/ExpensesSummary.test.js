import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test("Viewing 2 expenses with tottaling $5,100.00", () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={5100}/>);
    expect(wrapper).toMatchSnapshot();
});

test("Viewing 1 expenses with tottaling $100.00", () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={100}/>);
    expect(wrapper).toMatchSnapshot();
});