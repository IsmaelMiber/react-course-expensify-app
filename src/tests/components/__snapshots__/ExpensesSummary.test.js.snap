// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewing 1 expenses with tottaling $100.00 1`] = `
<p>
  Viewing 
  1
   expenses totalling 
  $1.00
</p>
`;

exports[`Viewing 2 expenses with tottaling $5,100.00 1`] = `
<p>
  Viewing 
  2
   expenses totalling 
  $51.00
</p>
`;
